core/src/main/java/io/jenkins/pluginhealth/scoring/probes/AbstractJavaImportsCheckProbe.java

```diff
@@ -0,0 +1,93 @@
+package io.jenkins.pluginhealth.scoring.probes;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import io.jenkins.pluginhealth.scoring.model.Plugin;
+import io.jenkins.pluginhealth.scoring.model.ProbeResult;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.util.CollectionUtils;
+
+/**
+ * The abstract probe checks for imports in {@code .java} files in a plugin.
+ */
+public abstract class AbstractJavaImportsCheckProbe extends Probe {
+    public static final int ORDER = SCMLinkValidationProbe.ORDER + 100;
+    private static final Logger LOGGER = LoggerFactory.getLogger(AbstractJavaImportsCheckProbe.class);
+
+    @Override
+    protected ProbeResult doApply(Plugin plugin, ProbeContext context) {
+        final Path scmRepository = context.getScmRepository();
+
+        /* The "maxDepth" is set to Integer.MAX_VALUE because a repository can have multiple modules with class files in it. We do not want to miss any ".java" file.  */
+        try (Stream<Path> paths = Files.find(scmRepository, Integer.MAX_VALUE, (path, $) ->
+            Files.isRegularFile(path) && path.getFileName().toString().endsWith(".java"))) {
+            Set<String> javaFilesWithDeprecatedImports = paths.map(this::getFileNameWithImports)
+                .filter(Objects::nonNull)
```

Comment:
From a design perspective, I think you can slightly change the methods you are using.

Here, your `getFileNameWithImports` is transparently responsible of 2 different things:

* Checking if a file contains some `import` statements,
* Return the name of the file if it contains the `import` statements.

In order to implement those 2 behaviors in only 1 method, you opted for some sort of convention: if the method returns a `String`, it means the file contains the import, and you now have its file name. But if the file does not contain the import, you return `null` and that `null` value is to be interpreted in "the import statement is not found in that file".

If you remember the article I shared with you some time ago about accronyms, there is one related to implementing methods which are responsible for one and only one thing.

Here, your method is doing 2 different things.

One way to handle this could be:

Change `getFileNameWithImports` to a method like `boolean containsImports(Path javaFile, List<String> imports)`. The code you wrote is correct, you just have to change the return statement to a boolean instead of returning `String`s.

Then, in your caller method, instead of implementing that "convention" of yours (filtering the `null` objects to remove the java classes which do not contain the imports (see how non intuitive that is?)), you can simply do the following:

`javaFiles.filter(javaFile -> containsImports(javaFile, getListOfImports()))`

This first part will simply filter the list of all java files, keeping only the ones with the detected imports.

Then, because you want to collect their file names, you just have to:

`.map(javaFile -> javaFile.getFileName().toString())`.

This way you don't rely on kind of an internal transparent convention to write your code, and your code becomes very explicit.

WDYT? Does it make sense?